components/snap_recovery/src/data_resolver.rs

```
@@ -0,0 +1,459 @@
+            let key = self.write_iter.key().to_vec();
+            self.write_iter.next().unwrap();
+            return Ok(Some((key, write)));
+        }
+        Ok(None)
+    }
+
+    fn scan_next_batch(&mut self) -> Result<Batch> {
+        let mut writes = Vec::with_capacity(self.batch_size_limit);
+        let mut has_more = true;
+
+        for _ in 0..self.batch_size_limit {
+            if let Some((key, write)) = self.next_batch()? {
+                let commit_ts = box_try!(Key::decode_ts_from(keys::origin_key(&key)));
+                if commit_ts > self.resolved_ts {
+                    writes.push((key, write));
+                }
+            } else {
+                has_more = false;
+                break;
+            }
+        }
+        Ok(Batch { writes, has_more })
+    }
+
+    // delete key.commit_ts > resolved-ts in write cf and default cf
+    fn batch_resolve_write(&mut self, wb: &mut RocksWriteBatchVec) -> Result<bool> {
+        let Batch { writes, has_more } = self.scan_next_batch()?;
+        if has_more && writes.is_empty() {
+            return Ok(has_more);
+        }
+
+        let batch = writes.clone();
+        let mut max_ts: TimeStamp = 0.into();
+        for (key, write) in writes {
+            let default_key = Key::from_encoded_slice(&key)
+                .truncate_ts()
+                .unwrap()
+                .append_ts(write.start_ts);
+            box_try!(wb.delete_cf(CF_WRITE, &key));
+            box_try!(wb.delete_cf(CF_DEFAULT, default_key.as_encoded()));
+
+            if write.start_ts > max_ts {
+                max_ts = write.start_ts;
+            }
+        }
+        info!(
+            "flush delete in write/default cf. delete key count {}",
+            batch.len()
+        );
+        let mut write_opts = WriteOptions::new();
+        write_opts.set_sync(true);
+        if let Err(e) = wb.write_opt(&write_opts) {
+            panic!(
```

We can return the error to the caller.

---

fixed